src/main/java/iterator/RichIterator.java

```diff
@@ -371,7 +612,8 @@ default <B> RichIterator<Pair<A, B>> zip(Iterator<B> that) {
      * @return the zipped iterator
      */
     default RichIterator<Pair<A, Integer>> zipWithIndex() {
-        throw new NotImplementedException();
+        AtomicInteger i = new AtomicInteger(-1);
```

Comment:
Atomic integers are not efficient. 
Try use the function `zip` and other function to make this work